### ISSUE TITLE
Blocked Hosts File

### ISSUE THREAD

**Issue by TiggaStyle** [arno-iptables-firewall/aif#31]
Hello,

i'am Running some VPS with Ubuntu 14.04. I've Installed this Firewall with a script to fill the blocked-host file. Everything runs Perfect.

So now i have tried it on Ubuntu 15.10 an 16.04 64bit Server.
Everything runs Perfect, till i filled up the blocked-host file.  I grep much sites like  blocklist.de all IPs or Ranges, so the File has 30K IPs. On Ubuntu 14.04 restart takes under a minute. In 15.10 or 16.04, the newest, the start/restart takes 15 Minutes. After reading the blocked-hosts file. It takes much time to start. The System is  on 100% cpu.

I Hope anyone understand me and can help

**Comment by abelbeck** [comment]
Interesting, I can't explain why 14.04 is so much faster.  Different memory, number of cores ?

Granted 30K blocked-host entries is a lot, and would be best served with using 'ipset' which AIF does not currently natively support.

I personally have 1K blocked-host entries and it takes 3 seconds to ```arno-iptables-firewall force-reload```  on a 1.80GHz Atom D525 using Linux 3.2.x .

**Comment by TiggaStyle** [comment]
Force reload loads only the Block hosts file New. Thats much faster on every System. But restart or Start takes time. So on Boot or Reboot. 
The machines are the Same. Only installed with 14.04 or 16.04.

I wonder why in 14.04 everything is okay and on newer System it takes so much time/Ressources

**Comment by abelbeck** [comment]
So the reloading of the Block hosts file is not the issue ? Same RAM and # of cores ?

Try manually ```arno-iptables-firewall restart``` and see where the delay is.

If that is as expected, then possibly Ubuntu added some sort of iptables save/restore feature that is taking the time.  If so possibly that can be disabled.

**Comment by abelbeck** [comment]
This may be of interest:
http://dev-notes.eu/2016/08/persistent-iptables-rules-in-ubuntu-16-04-xenial-xerus/

**Comment by TiggaStyle** [comment]
Ubuntu 14.04
/etc/init.d/arno-iptables-firewall restart
Sep 06 06:59:45 ** Restarting Arno's Iptables Firewall v2.0.1g-DEVEL **
Sep 06 07:00:07 ** All firewall rules applied **
Ubuntu 16.04
Sep 06 07:05:45 ** Restarting Arno's Iptables Firewall v2.0.1f **
Sep 06 07:12:44 ** All firewall rules applied **

And its not the Versions. If have tried it with a new Installation of Ubuntu and the RC1 of Arno.

 0%.........20%.........40%.........60%.........80%.........100%.........35658 host line(s) read
Using loglevel "info" for syslogd
Setting up firewall rules:
-------------------------------------------------------------------------------
Enabling setting the maximum packet size via MSS
<<<
After this it takes the time.

**Comment by TiggaStyle** [comment]
Thx to Everyone, i found a solution with a ipset blacklist. But it would be nice to have ipset direkt in arno.

**Comment by abelbeck** [comment]
@TiggaStyle thanks for the info, I don't understand the differences you report between Ubuntu versions.

You mentioned that reloading the 30K Block hosts file via ```force-reload``` was reasonably quick across all Ubuntu versions, but the fix was to use ```ipset``` which fixes standard start and restart.

Thanks again, we need to understand the real speed issue here.

**Comment by TiggaStyle** [comment]
I don't understand the difference, too.

restart or start takes 10 Minutes, force-reload goes fast under a minute. I think it must be a changes from Kernel or network stack. But i dont know.

I just noticed that the start from my server takes 10 minutes, and thats from arno with die blocklist.

**Comment by abelbeck** [comment]
Thanks again, glad you have a solution for now.

**Comment by abelbeck** [comment]
@TiggaStyle We just added ipset support to AIF.  Hopefully you have time to test it, you can get it from the master branch here or get it from the nightly package:
http://rocky.eld.leidenuniv.nl/arno-iptables-firewall/arno-iptables-firewall_nightly.tar.gz

I should show version 2.0.1g-RC2 .

You only need to add ```IPTABLES_IPSET=1``` to your firewall.conf configuration and it should work as it did before, but faster.  An additional benefit is during "force-reload" new connections will be allowed, which we had to block while adding rules before.

**Comment by TiggaStyle** [comment]
THX,
if got the clone from github, installed it with install.sh 
Config IPSET Log and Blocked-hosts
55k blocked-hosts with ipset testet on 2 Server under ubuntu 16.04, works perfect.
So i hope thats the block works.

Thanks for that :+1:

**Comment by abelbeck** [comment]
@TiggaStyle Thanks for testing.

**Comment by abelbeck** [comment]
@TiggaStyle We now have version 2.0.1g-RC4 available in the _nightly.tar.gz .  If you can test, it would be appreciated.

A new ```BLOCK_NETSET_DIR``` variable that uses ipset is now supported:
```
# Uncomment & specify here the location of the directory that contains *.netset
# files containing IP addresses and/or Networks (CIDR notation) to be BLOCKED.
# Files may only contain one entry per line, each IP/Net entry must start at
# the beginning of the line, other lines will be ignored.
# Each valid IP/Net entry must be followed by a new-line or a POSIX [[:space:]].
#
# File naming convention, use *v6.netset files for IPv6, all other *.netset
# files default to IPv4. Filenames are limited to 27-chars before the .netset .
# Optional whitelist.netset, whitelistv6.netset files may contain IP/Net entries
# that will not be BLOCKED by other *.netset files. Automatically by default,
# all Private (RFC1918), Link-Local and Multicast IP/Nets will be whitelisted.
#
# Note: This option depends on IPTABLES_IPSET being enabled.
# ------------------------------------------------------------------------------
#BLOCK_NETSET_DIR="/etc/arno-iptables-firewall/blocklists"
```
If you copy your ipset file (must be named with a .netset suffix) to your configured ```BLOCK_NETSET_DIR``` directory you should notice a greater than 10x speed improvement for ```force-reload``` .

You will want to disable ```BLOCK_HOSTS_FILE``` to get the full speed benefit.

Additionally, over on the AstLinux Project, here is a ```reload-blocklist-netset``` script I created to download the FireHOL firehol_level1.netset, it should work on a general distro.
https://svn.code.sf.net/p/astlinux/code/branches/1.0/package/arnofw/reload-blocklist-netset

**Comment by abelbeck** [comment]
Fixed in 2.0.1g
